src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb

```diff
@@ -559,28 +591,104 @@ Namespace Microsoft.CodeAnalysis.Semantics
                 Return Create(boundConversion.Operand)
             End If
 
-            Dim operand As Lazy(Of IOperation)
-            Dim methodSymbol As MethodSymbol
-
-            If (boundConversion.ConversionKind And VisualBasic.ConversionKind.UserDefined) = VisualBasic.ConversionKind.UserDefined Then
-                Dim userDefinedConversion As BoundUserDefinedConversion = DirectCast(boundConversion.Operand, BoundUserDefinedConversion)
-                methodSymbol = userDefinedConversion.Call.Method
-                operand = New Lazy(Of IOperation)(Function() Create(userDefinedConversion.Operand))
-            Else
-                methodSymbol = Nothing
-                operand = New Lazy(Of IOperation)(Function() Create(boundConversion.Operand))
+            Dim type As ITypeSymbol = boundConversion.Type
+            Dim constantValue As [Optional](Of Object) = ConvertToOptional(boundConversion.ConstantValueOpt)
+            Dim isImplicit As Boolean = boundConversion.WasCompilerGenerated OrElse Not boundConversion.ExplicitCastInCode
+
+            Dim conversionInformation = CreateConversionOperation(boundConversion.Operand, boundConversion.ConversionKind, boundConversion.Syntax)
+            Dim methodSymbol As MethodSymbol = conversionInformation.methodSymbol
+            Dim operand As Lazy(Of IOperation) = conversionInformation.operation
+            Dim isAddressOfDelegateCreation As Boolean = conversionInformation.isAddressOfDelegateCreation
+
+            If isAddressOfDelegateCreation OrElse
+               ((boundConversion.Operand.Kind = BoundKind.Lambda OrElse
+                 boundConversion.Operand.Kind = BoundKind.QueryLambda OrElse
+                 boundConversion.Operand.Kind = BoundKind.UnboundLambda) AndAlso
+                boundConversion.Type.IsDelegateType()) Then
```

Comment:
It looks like this condition is used in several places, consider extracting it into a helper method. Alternatively, the Type can be passed into CreateConversionOperation, which can perform all these checks and calculate isDelegateCreation rather than just calculating isAddressOfDelegateCreation.

Reply:
Refactored.

---
In reply to: [142806631](https://github.com/dotnet/roslyn/pull/22056#discussion_r142806631) [](ancestors = 142806631)